File: georocket-server/src/main/java/io/georocket/GeoRocket.java
package io.georocket;

import io.georocket.constants.ConfigConstants;
import io.georocket.http.Endpoint;
import io.georocket.index.IndexerVerticle;
import io.georocket.index.MetadataVerticle;
import io.georocket.tasks.TaskVerticle;
import io.georocket.util.FilteredServiceLoader;
import io.georocket.util.JsonUtils;
import io.georocket.util.SizeFormat;
import io.vertx.core.AbstractVerticle;
import io.vertx.core.DeploymentOptions;
import io.vertx.core.Future;
import io.vertx.core.Verticle;
import io.vertx.core.Vertx;
import io.vertx.core.http.HttpMethod;
import io.vertx.core.http.HttpServer;
import io.vertx.core.http.HttpServerOptions;
import io.vertx.core.json.DecodeException;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.core.logging.Logger;
import io.vertx.core.logging.LoggerFactory;
import io.vertx.core.net.PemKeyCertOptions;
import io.vertx.ext.web.Router;
import io.vertx.ext.web.handler.CorsHandler;
import io.vertx.rx.java.ObservableFuture;
import io.vertx.rx.java.RxHelper;
import org.apache.commons.io.FileUtils;
import org.apache.commons.io.IOUtils;
import org.jooq.lambda.Seq;
import org.yaml.snakeyaml.Yaml;
import rx.Completable;
import rx.Observable;
import rx.Single;
import rx.plugins.RxJavaHooks;

import java.io.File;
import java.io.IOException;
import java.lang.management.ManagementFactory;
import java.lang.management.MemoryMXBean;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * GeoRocket - A high-performance database for geospatial files
 * @author Michel Kraemer
 */
public class GeoRocket extends AbstractVerticle {
  private static Logger log = LoggerFactory.getLogger(GeoRocket.class);
  
  protected static File geoRocketHome;

  /**
   * Deploy a new verticle with the standard configuration of this instance
   * @param cls the class of the verticle class to deploy
   * @return a single that will carry the verticle's deployment id
   */
  protected Single<String> deployVerticle(Class<? extends Verticle> cls) {
    return Single.defer(() -> {
      ObservableFuture<String> observable = RxHelper.observableFuture();
      DeploymentOptions options = new DeploymentOptions().setConfig(config());
      vertx.deployVerticle(cls.getName(), options, observable.toHandler());
      return observable.toSingle();
    });
  }

  /**
   * Deploys all verticles from GeoRocket extensions (registered through Java
   * Service Provider Interface)
   * @return a completable that completes when all verticles have been deployed
   */
  protected Completable deployExtensionVerticles() {
    return Completable.defer(() -> {
      DeploymentOptions options = new DeploymentOptions().setConfig(config());
      return Observable.from(FilteredServiceLoader.load(ExtensionVerticle.class))
        .flatMap(verticle -> {
          ObservableFuture<String> observable = RxHelper.observableFuture();
          vertx.deployVerticle(verticle, options, observable.toHandler());
          return observable;
        })
        .toCompletable();
    });
  }

  /**
   * Deploy the indexer verticle
   * @return a single that will complete when the verticle was deployed
   * and will carry the verticle's deployment id
   */
  protected Single<String> deployIndexer() {
    return deployVerticle(IndexerVerticle.class);
  }

  /**
   * Deploy the importer verticle
   * @return a single that will complete when the verticle was deployed
   * and will carry the verticle's deployment id
   */
  protected Single<String> deployImporter() {
    return deployVerticle(ImporterVerticle.class);
  }

  /**
   * Deploy the metadata verticle
   * @return a single that will complete when the verticle was deployed
   * and will carry the verticle's deployment id
   */
  protected Single<String> deployMetadata() {
    return deployVerticle(MetadataVerticle.class);
  }

  /**
   * Deploy the task verticle
   * @return a single that will complete when the verticle was deployed
   * and will carry the verticle's deployment id
   */
  protected Single<String> deployTask() {
    return deployVerticle(TaskVerticle.class);
  }

  /**
   * Deploy the http server.
   * @return a single that will complete when the http server was started.
   */
  protected Single<HttpServer> deployHttpServer() {
    String host = config().getString(ConfigConstants.HOST, ConfigConstants.DEFAULT_HOST);
    int port = config().getInteger(ConfigConstants.PORT, ConfigConstants.DEFAULT_PORT);

    try {
      Router router = createRouter();
      HttpServerOptions serverOptions = createHttpServerOptions();
      HttpServer server = vertx.createHttpServer(serverOptions);
      ObservableFuture<HttpServer> observable = RxHelper.observableFuture();
      server.requestHandler(router::accept).listen(port, host, observable.toHandler());
      return observable.toSingle();
    } catch (Throwable t) {
      return Single.error(t);
    }
  }

  /**
   * Create and configure a {@link CorsHandler}
   * @return the {@link CorsHandler}
   */
  protected CorsHandler createCorsHandler() {
    String allowedOrigin = config().getString(
      ConfigConstants.HTTP_CORS_ALLOW_ORIGIN, "$."); // match nothing by default
    CorsHandler corsHandler = CorsHandler.create(allowedOrigin);

    // configure whether the Access-Control-Allow-Credentials should be returned
    if (config().getBoolean(ConfigConstants.HTTP_CORS_ALLOW_CREDENTIALS, false)) {
      corsHandler.allowCredentials(true);
    }

    // configured allowed headers
    Object allowHeaders = config().getValue(ConfigConstants.HTTP_CORS_ALLOW_HEADERS);
    if (allowHeaders instanceof String) {
      corsHandler.allowedHeader((String)allowHeaders);
    } else if (allowHeaders instanceof JsonArray) {
      corsHandler.allowedHeaders(Seq.seq((JsonArray)allowHeaders)
        .cast(String.class).toSet());
    } else if (allowHeaders != null) {
      throw new IllegalArgumentException(ConfigConstants.HTTP_CORS_ALLOW_HEADERS
        + " must either be a string or an array.");
    }

    // configured allowed methods
    Object allowMethods = config().getValue(ConfigConstants.HTTP_CORS_ALLOW_METHODS);
    if (allowMethods instanceof String) {
      corsHandler.allowedMethod(HttpMethod.valueOf((String)allowMethods));
    } else if (allowMethods instanceof JsonArray) {
      corsHandler.allowedMethods(Seq.seq((JsonArray)allowMethods)
        .cast(String.class).map(HttpMethod::valueOf).toSet());
    } else if (allowMethods != null) {
      throw new IllegalArgumentException(ConfigConstants.HTTP_CORS_ALLOW_METHODS
        + " must either be a string or an array.");
    }

    // configured exposed headers
    Object exposeHeaders = config().getValue(ConfigConstants.HTTP_CORS_EXPOSE_HEADERS);
    if (exposeHeaders instanceof String) {
      corsHandler.exposedHeader((String)exposeHeaders);
    } else if (exposeHeaders instanceof JsonArray) {
      corsHandler.exposedHeaders(Seq.seq((JsonArray)exposeHeaders)
        .cast(String.class).toSet());
    } else if (exposeHeaders != null) {
      throw new IllegalArgumentException(ConfigConstants.HTTP_CORS_EXPOSE_HEADERS
        + " must either be a string or an array.");
    }

    // configure max age in seconds
    int maxAge = config().getInteger(ConfigConstants.HTTP_CORS_MAX_AGE, -1);
    corsHandler.maxAgeSeconds(maxAge);

    return corsHandler;
  }
  
  /**
   * Create a {@link Router} and add routes for <code>/store/</code>
   * to it. Sub-classes may override if they want to add further routes
   * @return the created {@link Router}
   */
  protected Router createRouter() {
    Router router = Router.router(vertx);

    boolean corsEnable = config().getBoolean(ConfigConstants.HTTP_CORS_ENABLE, false);
    if (corsEnable) {
      router.route().handler(createCorsHandler());
    }

    for (Endpoint ep : FilteredServiceLoader.load(Endpoint.class)) {
      router.mountSubRouter(ep.getMountPoint(), ep.createRouter(vertx));
    }

    router.route().handler(ctx -> {
      String reason = "The endpoint " + ctx.request().path() + " does not exist";
      ctx.response()
        .setStatusCode(404)
        .end(ServerAPIException.toJson("endpoint_not_found", reason).toString());
    });
    
    return router;
  }

  /**
   * Create an {@link HttpServerOptions} object and modify it according to the
   * configuration. Sub-classes may override this method to further modify the
   * object.
   * @return the created {@link HttpServerOptions}
   */
  protected HttpServerOptions createHttpServerOptions() {
    boolean compress = config().getBoolean(ConfigConstants.HTTP_COMPRESS, true);

    HttpServerOptions serverOptions = new HttpServerOptions()
        .setCompressionSupported(compress);

    boolean ssl = config().getBoolean(ConfigConstants.HTTP_SSL, false);
    if (ssl) {
      serverOptions.setSsl(ssl);
      String certPath = config().getString(ConfigConstants.HTTP_CERT_PATH, null);
      String keyPath = config().getString(ConfigConstants.HTTP_KEY_PATH, null);
      PemKeyCertOptions pemKeyCertOptions = new PemKeyCertOptions()
          .setCertPath(certPath)
          .setKeyPath(keyPath);
      serverOptions.setPemKeyCertOptions(pemKeyCertOptions);
    }

    boolean alpn = config().getBoolean(ConfigConstants.HTTP_ALPN, false);
    if (alpn) {
      if (!ssl) {
        log.warn("ALPN is enabled but SSL is not! In order for ALPN to work " +
            "correctly, SSL is required.");
      }
      serverOptions.setUseAlpn(alpn);
    }

    return serverOptions;
  }
  
  @Override
  public void start(Future<Void> startFuture) {
    log.info("Launching GeoRocket " + getVersion() + " ...");

    deployExtensionVerticles()
      .doOnCompleted(() -> {
        vertx.eventBus().publish(ExtensionVerticle.EXTENSION_VERTICLE_ADDRESS,
            new JsonObject().put("type", ExtensionVerticle.MESSAGE_ON_INIT));
      })
      .andThen(deployTask()
        .flatMap(v -> deployIndexer())
        .flatMap(v -> deployImporter())
        .flatMap(v -> deployMetadata())
        .flatMap(v -> deployHttpServer())
      )
      .subscribe(id -> {
        vertx.eventBus().publish(ExtensionVerticle.EXTENSION_VERTICLE_ADDRESS,
            new JsonObject().put("type", ExtensionVerticle.MESSAGE_POST_INIT));
        log.info("GeoRocket launched successfully.");
        startFuture.complete();
      }, startFuture::fail);
  }
  
  /**
   * Replace configuration variables in a string
   * @param str the string
   * @return a copy of the given string with configuration variables replaced
   */
  private static String replaceConfVariables(String str) {
    return str.replace("$GEOROCKET_HOME", geoRocketHome.getAbsolutePath());
  }
  
  /**
   * Recursively replace configuration variables in an array
   * @param arr the array
   * @return a copy of the given array with configuration variables replaced
   */
  private static JsonArray replaceConfVariables(JsonArray arr) {
    JsonArray result = new JsonArray();
    for (Object o : arr) {
      if (o instanceof JsonObject) {
        replaceConfVariables((JsonObject)o);
      } else if (o instanceof JsonArray) {
        o = replaceConfVariables((JsonArray)o);
      } else if (o instanceof String) {
        o = replaceConfVariables((String)o);
      }
      result.add(o);
    }
    return result;
  }
  
  /**
   * Recursively replace configuration variables in an object
   * @param obj the object
   */
  private static void replaceConfVariables(JsonObject obj) {
    Set<String> keys = new HashSet<>(obj.getMap().keySet());
    for (String key : keys) {
      Object value = obj.getValue(key);
      if (value instanceof JsonObject) {
        replaceConfVariables((JsonObject)value);
      } else if (value instanceof JsonArray) {
        JsonArray arr = replaceConfVariables((JsonArray)value);
        obj.put(key, arr);
      } else if (value instanceof String) {
        String newValue = replaceConfVariables((String)value);
        obj.put(key, newValue);
      }
    }
  }
  
  /**
   * Set default configuration values
   * @param conf the current configuration
   */
  private static void setDefaultConf(JsonObject conf) {
    conf.put(ConfigConstants.HOME, "$GEOROCKET_HOME");
    if (!conf.containsKey(ConfigConstants.STORAGE_FILE_PATH)) {
      conf.put(ConfigConstants.STORAGE_FILE_PATH, "$GEOROCKET_HOME/storage");
    }
  }

  /**
   * Load the GeoRocket configuration
   * @return the configuration
   *
   * @throws IOException If the georocket home is invalid or the file could not be accessed
   * @throws DecodeException If the configuration could not be decoded from json
   */
  protected static JsonObject loadGeoRocketConfiguration() throws IOException, DecodeException {
    String geoRocketHomeStr = System.getenv("GEOROCKET_HOME");
    if (geoRocketHomeStr == null) {
      log.info("Environment variable GEOROCKET_HOME not set. Using current "
          + "working directory.");
      geoRocketHomeStr = new File(".").getAbsolutePath();
    }

    geoRocketHome = new File(geoRocketHomeStr).getCanonicalFile();

    log.info("Using GeoRocket home " + geoRocketHome);

    // load configuration file
    File confDir = new File(geoRocketHome, "conf");
    File confFile = new File(confDir, "georocketd.yaml");
    if (!confFile.exists()) {
      confFile = new File(confDir, "georocketd.yml");
      if (!confFile.exists()) {
        confFile = new File(confDir, "georocketd.json");
      }
    }
    String confFileStr = FileUtils.readFileToString(confFile, "UTF-8");
    JsonObject conf;
    if (confFile.getName().endsWith(".json")) {
      conf = new JsonObject(confFileStr);
    } else {
      Yaml yaml = new Yaml();
      @SuppressWarnings("unchecked")
      Map<String, Object> m = yaml.loadAs(confFileStr, Map.class);
      conf = JsonUtils.flatten(new JsonObject(m));
    }

    // set default configuration values
    setDefaultConf(conf);

    // replace variables in config
    replaceConfVariables(conf);
    
    overwriteWithEnvironmentVariables(conf);

    return conf;
  }

  /**
   * Match every environment variable against the config keys from
   * {{@link ConfigConstants#getConfigKeys()}} and save the found values using
   * the config key in the config object. The method is equivalent to calling
   * <code>overwriteWithEnvironmentVariables(conf, java.lang.System.getenv())</code>
   * @param conf the config object
   */
  private static void overwriteWithEnvironmentVariables(JsonObject conf) {
    overwriteWithEnvironmentVariables(conf, System.getenv());
  }

  /**
   * Match every environment variable against the config keys from
   * {{@link ConfigConstants#getConfigKeys()}} and save the found values using
   * the config key in the config object.
   * @param conf the config object
   * @param env the map with the environment variables
   */
  static void overwriteWithEnvironmentVariables(JsonObject conf,
    Map<String, String> env) {
    Map<String, String> names = ConfigConstants.getConfigKeys()
      .stream()
      .collect(Collectors.toMap(
        s -> s.toUpperCase().replace(".", "_"),
        Function.identity()
      ));
    env.forEach((key, val) -> {
      String name = names.get(key.toUpperCase());
      if (name != null) {
        Yaml yaml = new Yaml();
        Object newVal = yaml.load(val);
        conf.put(name, newVal);
      }
    });
  }

  /**
   * @return the tool's version string
   */
  public static String getVersion() {
    URL u = GeoRocket.class.getResource("version.dat");
    String version;
    try {
      version = IOUtils.toString(u, StandardCharsets.UTF_8);
    } catch (IOException e) {
      throw new RuntimeException("Could not read version information", e);
    }
    return version;
  }

  /**
   * Runs the server
   * @param args the command line arguments
   */
  public static void main(String[] args) {
    // print banner
    try {
      URL u = GeoRocket.class.getResource("georocket_banner.txt");
      String banner = IOUtils.toString(u, StandardCharsets.UTF_8);
      System.out.println(banner);
    } catch (IOException e) {
      // ignore
    }

    Vertx vertx = Vertx.vertx();

    // register schedulers that run Rx operations on the Vert.x event bus
    RxJavaHooks.setOnComputationScheduler(s -> RxHelper.scheduler(vertx));
    RxJavaHooks.setOnIOScheduler(s -> RxHelper.blockingScheduler(vertx));
    RxJavaHooks.setOnNewThreadScheduler(s -> RxHelper.scheduler(vertx));

    DeploymentOptions options = new DeploymentOptions();

    try {
      JsonObject conf = loadGeoRocketConfiguration();
      options.setConfig(conf);
    } catch (IOException ex) {
      log.fatal("Invalid georocket home", ex);
      System.exit(1);
    } catch (DecodeException ex) {
      log.fatal("Failed to decode the GeoRocket (JSON) configuration", ex);
      System.exit(1);
    }
    
    boolean logConfig = options.getConfig().getBoolean(
        ConfigConstants.LOG_CONFIG, false);
    if (logConfig) {
      log.info("Configuration:\n" + options.getConfig().encodePrettily());
    }

    // log memory info
    MemoryMXBean memoryMXBean = ManagementFactory.getMemoryMXBean();
    long memoryInit = memoryMXBean.getHeapMemoryUsage().getInit();
    long memoryMax = memoryMXBean.getHeapMemoryUsage().getMax();
    log.info("Initial heap size: " + SizeFormat.format(memoryInit) +
        ", max heap size: " + SizeFormat.format(memoryMax));

    // deploy main verticle
    vertx.deployVerticle(GeoRocket.class.getName(), options, ar -> {
        if (ar.failed()) {
          log.fatal("Could not deploy GeoRocket");
          ar.cause().printStackTrace();
          System.exit(1);
        }
      });
  }
}

File: georocket-server/src/main/kotlin/io/georocket/GeoRocket.kt
package io.georocket

import io.georocket.constants.ConfigConstants
import io.georocket.http.Endpoint
import io.georocket.http.GeneralEndpoint
import io.georocket.http.StoreEndpoint
import io.georocket.http.TaskEndpoint
import io.georocket.ogcapifeatures.OgcApiFeaturesEndpoint
import io.georocket.tasks.TaskVerticle
import io.georocket.util.FilteredServiceLoader
import io.georocket.util.JsonUtils
import io.georocket.util.SizeFormat
import io.vertx.core.DeploymentOptions
import io.vertx.core.Vertx
import io.vertx.core.http.HttpMethod
import io.vertx.core.http.HttpServerOptions
import io.vertx.core.json.DecodeException
import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject
import io.vertx.core.logging.LoggerFactory
import io.vertx.core.net.PemKeyCertOptions
import io.vertx.ext.web.Router
import io.vertx.ext.web.handler.CorsHandler
import io.vertx.kotlin.core.deployVerticleAwait
import io.vertx.kotlin.core.http.httpServerOptionsOf
import io.vertx.kotlin.core.http.listenAwait
import io.vertx.kotlin.coroutines.CoroutineVerticle
import org.apache.commons.io.FileUtils
import org.apache.commons.io.IOUtils
import org.yaml.snakeyaml.Yaml
import java.io.File
import java.io.IOException
import java.lang.management.ManagementFactory
import java.nio.charset.StandardCharsets
import kotlin.system.exitProcess

private val log = LoggerFactory.getLogger(GeoRocket::class.java)
private lateinit var geoRocketHome: File

/**
 * GeoRocket - A high-performance database for geospatial files
 * @author Michel Kraemer
 */
class GeoRocket : CoroutineVerticle() {
  private val endpoints = mutableListOf<Endpoint>()

  /**
   * Deploy the http server.
   * @return a single that will complete when the http server was started.
   */
  private suspend fun deployHttpServer() {
    val host = config.getString(ConfigConstants.HOST, ConfigConstants.DEFAULT_HOST)
    val port = config.getInteger(ConfigConstants.PORT, ConfigConstants.DEFAULT_PORT)
    val router = createRouter()
    val serverOptions = createHttpServerOptions()
    val server = vertx.createHttpServer(serverOptions)
    server.requestHandler(router).listenAwait(port, host)
  }

  /**
   * Create and configure a [CorsHandler]
   * @return the [CorsHandler]
   */
  private fun createCorsHandler(): CorsHandler {
    val allowedOrigin = config.getString(
        ConfigConstants.HTTP_CORS_ALLOW_ORIGIN, "$.") // match nothing by default
    val corsHandler = CorsHandler.create(allowedOrigin)

    // configure whether the Access-Control-Allow-Credentials should be returned
    if (config.getBoolean(ConfigConstants.HTTP_CORS_ALLOW_CREDENTIALS, false)) {
      corsHandler.allowCredentials(true)
    }

    // configured allowed headers
    val allowHeaders = config.getValue(ConfigConstants.HTTP_CORS_ALLOW_HEADERS)
    when {
      allowHeaders is String ->
        corsHandler.allowedHeader(allowHeaders)
      allowHeaders is JsonArray ->
        corsHandler.allowedHeaders(allowHeaders.map { it as String }.toSet())
      allowHeaders != null ->
        throw IllegalArgumentException(ConfigConstants.HTTP_CORS_ALLOW_HEADERS +
            " must either be a string or an array.")
    }

    // configured allowed methods
    val allowMethods = config.getValue(ConfigConstants.HTTP_CORS_ALLOW_METHODS)
    when {
      allowMethods is String ->
        corsHandler.allowedMethod(HttpMethod.valueOf(allowMethods))
      allowMethods is JsonArray ->
        corsHandler.allowedMethods(allowMethods.map { it as String }
            .map { HttpMethod.valueOf(it) }.toSet())
      allowMethods != null ->
        throw IllegalArgumentException(ConfigConstants.HTTP_CORS_ALLOW_METHODS +
            " must either be a string or an array.")
    }

    // configured exposed headers
    val exposeHeaders = config.getValue(ConfigConstants.HTTP_CORS_EXPOSE_HEADERS)
    when {
      exposeHeaders is String ->
        corsHandler.exposedHeader(exposeHeaders)
      exposeHeaders is JsonArray ->
        corsHandler.exposedHeaders(exposeHeaders.map { it as String}.toSet())
      exposeHeaders != null ->
        throw IllegalArgumentException(ConfigConstants.HTTP_CORS_EXPOSE_HEADERS +
            " must either be a string or an array.")
    }

    // configure max age in seconds
    val maxAge = config.getInteger(ConfigConstants.HTTP_CORS_MAX_AGE, -1)
    corsHandler.maxAgeSeconds(maxAge)
    return corsHandler
  }

  /**
   * Create a [Router] and add routes for `/store/` to it
   */
  private suspend fun createRouter(): Router {
    val router = Router.router(vertx)

    val corsEnable = config.getBoolean(ConfigConstants.HTTP_CORS_ENABLE, false)
    if (corsEnable) {
      router.route().handler(createCorsHandler())
    }

    val ge = GeneralEndpoint(vertx)
    router.mountSubRouter("/", ge.createRouter())

    val se = StoreEndpoint(coroutineContext, vertx)
    router.mountSubRouter("/store", se.createRouter())

    val te = TaskEndpoint(coroutineContext, vertx)
    router.mountSubRouter("/tasks", te.createRouter())

    val ogc = OgcApiFeaturesEndpoint(coroutineContext, vertx)
    router.mountSubRouter("/ogcapifeatures", ogc.createRouter())

    router.route().handler { ctx ->
      val reason = "The endpoint ${ctx.request().path()} does not exist"
      ctx.response()
          .setStatusCode(404)
          .end(ServerAPIException.toJson("endpoint_not_found", reason).toString())
    }

    endpoints.add(ge)
    endpoints.add(se)
    endpoints.add(te)
    endpoints.add(ogc)

    return router
  }

  /**
   * Create an [HttpServerOptions] object and modify it according to the
   * configuration
   */
  private fun createHttpServerOptions(): HttpServerOptions {
    val compress = config.getBoolean(ConfigConstants.HTTP_COMPRESS, true)
    val serverOptions = httpServerOptionsOf(compressionSupported = compress)

    val ssl = config.getBoolean(ConfigConstants.HTTP_SSL, false)
    if (ssl) {
      serverOptions.isSsl = ssl
      val certPath = config.getString(ConfigConstants.HTTP_CERT_PATH, null)
      val keyPath = config.getString(ConfigConstants.HTTP_KEY_PATH, null)
      val pemKeyCertOptions = PemKeyCertOptions()
          .setCertPath(certPath)
          .setKeyPath(keyPath)
      serverOptions.pemKeyCertOptions = pemKeyCertOptions
    }

    val alpn = config.getBoolean(ConfigConstants.HTTP_ALPN, false)
    if (alpn) {
      if (!ssl) {
        log.warn("ALPN is enabled but SSL is not! In order for ALPN to work " +
            "correctly, SSL is required.")
      }
      serverOptions.isUseAlpn = alpn
    }

    return serverOptions
  }

  override suspend fun start() {
    log.info("Launching GeoRocket ${getVersion()} ...")

    val options = DeploymentOptions().setConfig(config)

    // deploy extension verticles
    for (verticle in FilteredServiceLoader.load(ExtensionVerticle::class.java)) {
      vertx.deployVerticleAwait(verticle, options)
    }

    vertx.eventBus().publish(ExtensionVerticle.EXTENSION_VERTICLE_ADDRESS,
        JsonObject().put("type", ExtensionVerticle.MESSAGE_ON_INIT))

    // deploy other verticles
    vertx.deployVerticleAwait(TaskVerticle(), options)
    vertx.deployVerticleAwait(ImporterVerticle(), options)

    // deploy HTTP server
    deployHttpServer()

    vertx.eventBus().publish(ExtensionVerticle.EXTENSION_VERTICLE_ADDRESS,
        JsonObject().put("type", ExtensionVerticle.MESSAGE_POST_INIT))

    log.info("GeoRocket launched successfully.")
  }

  override suspend fun stop() {
    endpoints.forEach { it.close() }
  }

  /**
   * Return the tool's version string
   */
  private fun getVersion(): String {
    val u = GeoRocket::class.java.getResource("version.dat")
    return try {
      IOUtils.toString(u, StandardCharsets.UTF_8)
    } catch (e: IOException) {
      throw RuntimeException("Could not read version information", e)
    }
  }
}

/**
 * Replace configuration variables in a string
 */
private fun replaceConfVariables(str: String): String {
  return str.replace("\$GEOROCKET_HOME", geoRocketHome.absolutePath)
}

/**
 * Recursively replace configuration variables in an array
 */
private fun replaceConfVariables(arr: JsonArray): JsonArray {
  val result = JsonArray()
  for (o in arr) {
    val ro = when (o) {
      is JsonObject -> replaceConfVariables(o)
      is JsonArray -> replaceConfVariables(o)
      is String -> replaceConfVariables(o)
      else -> o
    }
    result.add(ro)
  }
  return result
}

/**
 * Recursively replace configuration variables in an object
 */
private fun replaceConfVariables(obj: JsonObject): JsonObject {
  val result = obj.copy()

  for (key in result.map.keys) {
    when (val value = result.getValue(key)) {
      is JsonObject -> result.put(key, replaceConfVariables(value))
      is JsonArray -> result.put(key, replaceConfVariables(value))
      is String -> result.put(key, replaceConfVariables(value))
    }
  }

  return result
}

/**
 * Set default configuration values
 */
private fun setDefaultConf(conf: JsonObject) {
  conf.put(ConfigConstants.HOME, "\$GEOROCKET_HOME")
  if (!conf.containsKey(ConfigConstants.STORAGE_FILE_PATH)) {
    conf.put(ConfigConstants.STORAGE_FILE_PATH, "\$GEOROCKET_HOME/storage")
  }
}

/**
 * Load the GeoRocket configuration
 */
private fun loadGeoRocketConfiguration(): JsonObject {
  var geoRocketHomeStr = System.getenv("GEOROCKET_HOME")
  if (geoRocketHomeStr == null) {
    log.info("Environment variable GEOROCKET_HOME not set. Using current " +
        "working directory.")
    geoRocketHomeStr = File(".").absolutePath
  }

  geoRocketHome = File(geoRocketHomeStr).canonicalFile
  log.info("Using GeoRocket home $geoRocketHome")

  // load configuration file
  val confDir = File(geoRocketHome, "conf")
  var confFile = File(confDir, "georocketd.yaml")
  if (!confFile.exists()) {
    confFile = File(confDir, "georocketd.yml")
    if (!confFile.exists()) {
      confFile = File(confDir, "georocketd.json")
    }
  }

  val confFileStr = FileUtils.readFileToString(confFile, "UTF-8")
  var conf: JsonObject = if (confFile.name.endsWith(".json")) {
    JsonObject(confFileStr)
  } else {
    val yaml = Yaml()
    @Suppress("UNCHECKED_CAST")
    val m = yaml.loadAs(confFileStr, HashMap::class.java) as Map<String, Any>
    JsonUtils.flatten(JsonObject(m))
  }

  // set default configuration values
  setDefaultConf(conf)

  // replace variables in config
  conf = replaceConfVariables(conf)
  overwriteWithEnvironmentVariables(conf)

  return conf
}

/**
 * Match every environment variable against the config keys from
 * [ConfigConstants.getConfigKeys] and save the found values using
 * the config key in the config object. The method is equivalent to calling
 * [overwriteWithEnvironmentVariables]
 */
private fun overwriteWithEnvironmentVariables(conf: JsonObject) {
  overwriteWithEnvironmentVariables(conf, System.getenv())
}

/**
 * Match every environment variable against the config keys from
 * [ConfigConstants.getConfigKeys] and save the found values using
 * the config key in the config object.
 */
fun overwriteWithEnvironmentVariables(conf: JsonObject, env: Map<String, String>) {
  val names = ConfigConstants.getConfigKeys()
    .associateBy { s -> s.uppercase().replace(".", "_") }
  env.forEach { (key, v) ->
    val name = names[key.uppercase()]
    if (name != null) {
      val yaml = Yaml()
      val newVal = yaml.load<Any>(v)
      conf.put(name, newVal)
    }
  }
}

/**
 * Run the server
 */
suspend fun main() {
  // print banner
  val banner = GeoRocket::class.java.getResource("georocket_banner.txt")!!.readText()
  println(banner)

  val vertx = Vertx.vertx()

  val options = DeploymentOptions()
  try {
    val conf = loadGeoRocketConfiguration()
    options.config = conf
  } catch (ex: IOException) {
    log.fatal("Invalid georocket home", ex)
    exitProcess(1)
  } catch (ex: DecodeException) {
    log.fatal("Failed to decode the GeoRocket (JSON) configuration", ex)
    exitProcess(1)
  }
  val logConfig = options.config.getBoolean(
      ConfigConstants.LOG_CONFIG, false)
  if (logConfig) {
    log.info("""
      Configuration:
      ${options.config.encodePrettily()}
      """.trimIndent())
  }

  // log memory info
  val memoryMXBean = ManagementFactory.getMemoryMXBean()
  val memoryInit = memoryMXBean.heapMemoryUsage.init
  val memoryMax = memoryMXBean.heapMemoryUsage.max
  log.info("Initial heap size: ${SizeFormat.format(memoryInit)}, " +
      "max heap size: ${SizeFormat.format(memoryMax)}")

  // deploy main verticle
  try {
    vertx.deployVerticleAwait(GeoRocket::class.java.name, options)
  } catch (t: Throwable) {
    log.fatal("Could not deploy GeoRocket")
    t.printStackTrace()
    exitProcess(1)
  }
}
